core/templates/pages/diagnostic-test-player-page/diagnostic-test.model.ts

```diff
@@ -0,0 +1,259 @@
+      }).length;
+
+      // The length of successor topic IDs that are currently present in
+      // the eligible topic IDs list.
+      let lengthOfEligibleSuccessorTopicIds = successors.filter((topic) => {
+        return (this._eligibleTopicIds.indexOf(topic) !== -1);
+      }).length;
+
+      // Considering a given topic, the assured number of topics that will be
+      // removed from the eligible topic IDs after testing, is the minimum
+      // between its ancestors and successors.
+      topicIdToLengthOfRelatedTopicIds[topicId] = Math.min(
+        lengthOfEligibleAncestorTopicIds, lengthOfEligibleSuccessorTopicIds);
+    }
+
+    // Among all the eligible topics, the topic with the maximum value for
+    // min(ancestors, successors) should be selected for testing. The
+    // maximum value helps us to reach the result faster which helps in quicker
+    // recommendations.
+    this._currentTopicId = Object.keys(topicIdToLengthOfRelatedTopicIds).reduce(
+      (item1, item2) => {
+        return (
+          topicIdToLengthOfRelatedTopicIds[item1] >=
+          topicIdToLengthOfRelatedTopicIds[item2] ? item1 : item2);
+      });
+    return this._currentTopicId;
+  }
+
+  recordTopicPassed(): void {
+    // If a topic is passed, that means we don't need to test any of the
+    // ancestor topics.
+    let ancestors = this._topicIdToAncestorTopicIds[this._currentTopicId];
+
+    let topicIdsToRemoveFromEligibleList: string[] = [];
+    topicIdsToRemoveFromEligibleList = (
+      topicIdsToRemoveFromEligibleList.concat(ancestors));
+    topicIdsToRemoveFromEligibleList.push(this._currentTopicId);
+
+    this._eligibleTopicIds = this._eligibleTopicIds.filter((topicId) => {
+      return (topicIdsToRemoveFromEligibleList.indexOf(topicId) === -1);
+    });
+  }
+
+  recordTopicFailed(): void {
+    // If a topic is failed, that means we don't need to test any of the
+    // successor topics.
+    let successors = this._topicIdToSuccessorTopicIds[this._currentTopicId];
+    this._failedTopicIds.push(this._currentTopicId);
+
+    let topicIdsToRemoveFromEligibleList: string[] = [];
+    topicIdsToRemoveFromEligibleList = (
+      topicIdsToRemoveFromEligibleList.concat(successors));
+    topicIdsToRemoveFromEligibleList.push(this._currentTopicId);
+
+    this._eligibleTopicIds = this._eligibleTopicIds.filter((topicId) => {
+      return (topicIdsToRemoveFromEligibleList.indexOf(topicId) === -1);
+    });
+  }
+
+  isTestFinished(): boolean {
+    // If the eligible topic IDs list is empty then the test should terminate.
+    if (this._eligibleTopicIds.length === 0) {
+      return true;
+    }
+
+    // If the eligible topic IDs list is non-empty but the learner has attempted
+    // the maximum allowed questions for the diagnostic test, then the test
+    // should terminate.
+    if (
+      this._eligibleTopicIds.length > 0 &&
+        this._totalNumberOfAttemptedQuestions >=
```

Comment:
I think you can deindent this line by 2 (but OK to keep the next line indented).

Reply:
Removed